physics/checkpointer.hpp

```diff
@@ -25,62 +26,68 @@ using quantities::Time;
-  // It must return true iff the given message actually contained a checkpoint.
-  // If it returns false, it must leave the object in a state corresponding to
-  // its default initialization.  This function is expected to capture the
-  // object being deserialized.
-  using Reader = std::function<bool(Message const&)>;
-
   // A function that writes an object to a checkpoint.  This function is
   // expected to capture the object being serialized.
-  using Writer = std::function<void(not_null<Message*>)>;
+  using Writer = std::function<void(not_null<typename Message::Checkpoint*>)>;
+
+  // A function that reconstructs an object from a checkpoint as a side effect.
+  // This function is expected to capture the object being deserialized.
+  using Reader = std::function<void(typename Message::Checkpoint const&)>;
 
-  Checkpointer(Reader reader, Writer writer);
+  Checkpointer(Writer writer, Reader reader);
+
+  // Returns the oldest checkpoint in this object, or +∞ if no checkpoint was
+  // ever created.
+  Instant oldest_checkpoint() const EXCLUDES(lock_);
 
   // Creates a checkpoint at time |t|, which will be used to recreate the
   // timeline after |t|.  The checkpoint is constructed by calling the |Writer|
   // passed at construction.
-  void CreateUnconditionally(Instant const& t) EXCLUDES(lock_);
+  void WriteToCheckpoint(Instant const& t) EXCLUDES(lock_);
 
   // Same as above, but a checkpoint is only created if one was not created
-  // recently, as specified by |max_time_between_checkpoints|.
-  bool CreateIfNeeded(Instant const& t,
-                      Time const& max_time_between_checkpoints) EXCLUDES(lock_);
-
-  // If there exists a checkpoint, writes the oldest checkpoint to the |message|
-  // using protocol buffer merging and returns its time.  Otherwise returns +∞.
-  // The time returned by this function should be serialized and passed to
-  // |ReadFromMessage| when deserializing to ensure that checkpoints are
-  // preserved across serialization/deserialization cycles.
-  Instant WriteToMessage(not_null<Message*> message) const EXCLUDES(lock_);
-
-  // Clears all the checkpoints in this checkpointer, and calls the |Reader|
-  // passed at construction to reconstruct the object from |message|.  If the
-  // |Reader| returns true (i.e., there was a checkpoint in the |message|),
-  // creates a new checkpoint in this checkpointer at the given time.
-  void ReadFromMessage(Instant const& t,
-                       Message const& message) EXCLUDES(lock_);
+  // recently, as specified by |max_time_between_checkpoints|.  Returns true iff
+  // a new checkpoint was created.
+  bool WriteToCheckpointIfNeeded(Instant const& t,
+                                 Time const& max_time_between_checkpoints)
+      EXCLUDES(lock_);
+
+  // Caller the |Reader| passed at construction to reconstruct an object using
+  // the oldest checkpoint.  Returns true iff this object contains at least one
+  // checkpoint.
```

Comment:
Could we return an erroneous status, and check IsFinite(oldest_checkpoint()) beforehand if needed? I find calls to `ReadFromOldestCheckpoint` that could actually do nothing a bit unreadable.

Reply:
It really should not happen, at the time of deserialization, that we don't have a checkpoint (it can happen e.g., when the checkpointer was just created) because we create one when serializing.  So I am changing this method to `ReadFromOldestCheckpointOrDie`.  This requires changes to tests that were actually not putting the object in a consistent state.

Reply:
I don’t think we should call our functions `OrDie`; doing anything with the `OldestCheckpoint` requires that the set of checkpoints be nonempty, so this is clearly a requirement from the name, and we check our requirements everywhere without calling everything `OrDie`.